### PR TITLE
[LS1-3693] Upgrade to .NET 8 and modernize dependencies

### PR DESCRIPTION
**Pull request by rustyclarksonfm**
- Upgrades target framework from netcoreapp3.1 to net8.0 and Docker base images from 3.1 to 8.0
- Replaces GitVersionTask (5.5.1, ARM64-incompatible) with GitVersion.MsBuild 6.7.0
- Upgrades FM.LiveSwitch SDK to 1.22.2 (pinned; 1.23.0+ requires a macOS native OpenSSL binary not yet included in the NuGet packages — tracked in LS1-3694)
- Replaces Newtonsoft.Json with System.Text.Json source generation (HammerJsonContext) for trim compatibility
- Rewrites TryGetOptions reflection with [DynamicallyAccessedMembers] to eliminate IL trim warnings
- Adds TrimmerRootDescriptor.xml to root the assembly and prevent trimmer from removing reflection-invoked constructors
- Suppresses only genuinely third-party warnings: IL2104 (CommandLineParser), NETSDK1206 (SkiaSharp via FM.LiveSwitch SDK)
- Fixes README dotnet publish/dotnet build commands to specify the project file explicitly, avoiding NETSDK1194

---

> [!NOTE]
> **Medium Risk**
> Migrating scan JSON and bumping the LiveSwitch SDK can change API parsing and media-test behavior; trimmed single-file publishes add reflection/serialization edge cases despite trimmer roots.
> 
> **Overview**
> Upgrades the Hammer CLI from **.NET Core 3.1** to **.NET 8** across the project, **Dockerfile**, and **GitHub Actions** (checkout/setup-dotnet/upload-artifact bumped to v4, SDK/runtime **8.0**).
> 
> **Dependencies:** Replaces **GitVersionTask** with **GitVersion.MsBuild** 6.7.0, bumps **CommandLineParser** and **FM.LiveSwitch** packages to **1.22.2**, and **drops Newtonsoft.Json** entirely.
> 
> **JSON:** Scan output and admin API responses now use **System.Text.Json** with a source-generated **`HammerJsonContext`** (including **`ScanTestOutput`** and explicit **`ipAddresses`** mapping on **`MediaServerInfo`**). Scan stdout shape stays **`failed`/`expiring`** via camelCase naming.
> 
> **Trimming / reflection:** Adds **`TrimmerRootDescriptor.xml`** for CLI option types, **`[DynamicallyAccessedMembers]`** on env-var option discovery, and targeted **`NoWarn`** for third-party trim/RID noise. **`Program`** skips **`ReadKey`** when stdin is redirected, maps **`JsonException`** to exit code 8, and logs exception type names.
> 
> **Docs / hygiene:** README requires .NET 8 and documents explicit **`.csproj`** paths for publish/build; **`.gitignore`** ignores **`osx/`**, **`linux/`**, **`win/`** publish folders; Codacy badge removed from README.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 908c51da1aab730be3e64c3d464a1c0f45f8b363. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by andriy-baranskyy — DISMISSED** [review]
1. Docker cloud
2. I think the claude comments are minor, ok to ignore.

**Comment by andriy-baranskyy** [comment]
Some comments by claude are worthwhile to address

**Comment by andriy-baranskyy** [comment]
## Comprehensive review — multi-agent (code, types, comments, error handling)

Thanks for the upgrade — the structural choices (STJ source-gen, explicit `[DynamicallyAccessedMembers]`, scoped suppressions) are right. Below are findings that surfaced across the review aspects, ordered by severity. The biggest theme: **silent behavior divergence between Newtonsoft and System.Text.Json on the deserialization path** (case sensitivity, null handling, naming policy on acronyms).

### Critical (must-fix before merge)

(1) **Case-sensitive deserialization — silent data loss vs. Newtonsoft.** `HammerJsonContext` does not set `PropertyNameCaseInsensitive = true`. Newtonsoft was case-insensitive by default; STJ is not. The fact that `[JsonPropertyName("ipAddresses")]` was already needed at `MediaServerInfo.cs:21` is evidence we already hit one mismatch. Any field on `MediaServerInfo`, `DeploymentConfig`, `CapacityThresholds` whose wire casing doesn't exactly match the camelCase policy will bind to default (null/0/false) with no exception. **Recommend:** add `PropertyNameCaseInsensitive = true` to `JsonSourceGenerationOptions`, and/or capture a real response from `GET v1/mediaservers` and `GET v2/DeploymentConfig` and verify each property.

(2) **`Program.Run` has no fallback `catch` — new `JsonException` paths crash the process.** `Run` returns only inside the typed catches. Anything else (notably `System.Text.Json.JsonException` thrown from the new `JsonSerializer.Deserialize` calls in `ScanTest.cs:258` and `:270`) propagates through `GetAwaiter().GetResult()` in `Main` and exits with a raw stack trace instead of the consistent `LogException` + exit-code contract. **Recommend:** add a `catch (JsonException ex)` with endpoint context, plus a final `catch (Exception ex) { LogException(ex); return <code>; }`.

(3) **`JsonSerializer.Deserialize` results not null-checked.** `ScanTest.cs:258` and `:270`: a JSON body of literal `null` returns `null`. The next-line `.Where(...)` and `deploymentConfig.CapacityThresholds` (`ScanTest.cs:203`) will throw `NullReferenceException`, which is not caught (see item 2). Pre-existing pattern, but the migration is the right time to fix it. **Recommend:** null-check both deserialize results and throw a typed exception with the endpoint URL and a snippet of the response.

### Important

(4) **`DefaultIgnoreCondition.WhenWritingNull` only affects serialization.** Newtonsoft's `NullValueHandling.Ignore` ignored nulls on **both** ends. Now an explicit `"foo": null` from the API will overwrite defaults — and for non-nullable value types on `MediaServerInfo` (`Available`, `Active`, `OverCapacity`, `UsedCapacity`, `Draining`, `CoreCount`) it raises `JsonException` instead of silently using defaults. Combined with item 2 this becomes a hard crash.

(5) **Naming-policy gotcha on acronyms (`PublicIPAddress`, `Mcu*` thresholds).** `JsonNamingPolicy.CamelCase` does not collapse leading runs of uppercase: `IPAddresses` → `iPAddresses` (which is exactly why the explicit `[JsonPropertyName("ipAddresses")]` exists). The same pattern applies to `PublicIPAddress` (policy yields `publicIPAddress`, API likely sends `publicIpAddress`) and to `CapacityThresholds.Mcu*ConnectionsPerCpuThreshold`. If Newtonsoft only worked here because of its case-insensitive read, these will silently start coming back as `null`/`0` after the upgrade. Verify against a real response and add `[JsonPropertyName(...)]` where needed.

(6) **csproj suppression comments are technically inaccurate.**
   - `FM.LiveSwitch.Hammer.csproj:7-8`: `IL2104` is the trimmer's per-assembly **summary** warning — it's not specific to CommandLineParser and not specifically about reflection. A future maintainer who upgrades CommandLineParser may delete the suppression assuming it's resolved, only to be blocked by IL2104 from another assembly (e.g. FM.LiveSwitch SDK).
   - `NETSDK1206` is the generic "version-specific RIDs" warning, not Tizen-specific. Recommend dropping the "Tizen" attribution and just noting it stems from legacy version-specific RIDs in transitive native packages.

(7) **Missing WHY for the FM.LiveSwitch 1.22.2 pin.** The PR description explains it; the csproj does not. Without an inline `<!-- ... -->` next to the three `FM.LiveSwitch*` `PackageReference`s, the next routine dependency bump will move to 1.25.x and reintroduce the regression. One line documenting the BouncyCastle / trim constraint is enough.

### Minor / informational

(8) **`TrimmerRootDescriptor.xml` with `preserve="all"` neutralizes most trim-correctness signal.** Pragmatic, but please add a one-line XML comment explaining *why* (CommandLineParser reflects over verb/option types — `ClusterTestOptions`, `LoadTestOptions`, `ScanTestOptions` and their `[Option]` properties — and would otherwise be trimmed). This couples the descriptor to the IL2104 suppression and prevents a future "tightening" pass from breaking arg parsing at runtime.

(9) **`MediaServerInfo.IPAddresses` — explain the explicit attribute inline.** A one-liner ("CamelCase policy emits `iPAddresses` for consecutive capitals; force lowercase") prevents a future cleanup from deleting the `[JsonPropertyName]`.

(10) **Background `Task.Run(Console.ReadKey)` swallows `InvalidOperationException` under redirected stdin** (`Program.cs:55-65`). Pre-existing, but interacts badly with the new "pipe JSON output" workflow this PR formalizes. Wrap in try/catch and fall back when `Console.IsInputRedirected`.

(11) **`LogException` drops type and stack** (`Program.cs:112-115`). With JSON parsing now in the hot path, a malformed response will surface as an opaque `'<' is an invalid start of a value…` with no endpoint. Consider including `ex.GetType().Name` and `ex.ToString()` at a verbose level.

(12) **`HammerJsonContext` has no doc-comment.** Adding "Source-generated JSON context required for PublishTrimmed; add a `[JsonSerializable(typeof(T))]` line for any new top-level (de)serialized type" prevents the silent runtime failure mode for future contributors adding a new DTO.

(13) **`ScanTestOutput` uses `IEnumerable<T>` with `init` setters.** Fine for the current write-only path; STJ source-gen does **not** support `IEnumerable<T>` for deserialization, so if this type is ever round-tripped this will fail. Consider `IReadOnlyList<T>` to lock intent.

### Confirmed-safe (called out so reviewers don't re-question them)

- `JsonStringEnumConverter<T>` (generic) is .NET 8+ and the correct source-gen + trim-compatible form. Behavior preserved (`DisplayAttribute` was never honored by Newtonsoft's `StringEnumConverter` either — only `EnumMemberAttribute`).
- `[DynamicallyAccessedMembers(PublicProperties)]` on `TryGetOptionsForType<T>` is sufficient for `GetProperties()` + reading property-level attributes; concrete type arguments (`ClusterTestOptions`/`LoadTestOptions`/`ScanTestOptions`) are reachable.
- Refactor from `Assembly.GetTypes()` to three explicit calls preserves first-match semantics (only those three types carry `[Verb]`).
- Type registration in `HammerJsonContext` is complete for the current call sites; transitively reachable types (`ScanTestMediaServerResult`, `ScanTestResult`, `ScanTestScenarioResult`, enums, `CapacityThresholds`) are auto-generated.
- `TrimmerRootDescriptor.xml` `assembly fullname="lshammer"` matches `<AssemblyName>lshammer</AssemblyName>`.
- README `dotnet publish/build` commands match the workflow invocations; CI bumps (`actions/checkout@v4`, `setup-dotnet@v4`, `upload-artifact@v4`, `dotnet-version: 8.0.x`) are coherent.

### Recommended action

1. Fix items 1 (`PropertyNameCaseInsensitive`), 2 (catch fallback), 3 (null-check) — these are blockers.
2. Verify items 4 and 5 against a real LiveSwitch API response before merging.
3. Tighten the comments in items 6/7/8/9/12.
4. Re-run review after fixes.

**Review by AnkitIsLive — COMMENTED** [review]
Looks good to me. Once andriy's changes are addressed

File: src/FM.LiveSwitch.Hammer/ScanTestState.cs [hunk 1]
```diff
@@ -1,5 +1,8 @@
-﻿namespace FM.LiveSwitch.Hammer
+﻿using System.Text.Json.Serialization;
+
+namespace FM.LiveSwitch.Hammer
 {
+    [JsonConverter(typeof(JsonStringEnumConverter<ScanTestState>))]
```
Comment (andriy-baranskyy) [inline src/FM.LiveSwitch.Hammer/ScanTestState.cs]: Wire-format regression: enum JSON values changed casing.

`HammerJsonContext` sets `PropertyNamingPolicy = JsonKnownNamingPolicy.CamelCase`, but `[JsonConverter(typeof(JsonStringEnumConverter<T>))]` constructed without an explicit `namingPolicy` argument **ignores** the context's `PropertyNamingPolicy` and emits enum members **verbatim** (member name).

- Before this PR: Newtonsoft `StringEnumConverter` + `CamelCaseNamingStrategy` emitted `"pass"`, `"fail"`, `"host"`, `"stun"`, `"turnUdp"`, `"turnTcp"`, `"turns"`.
- After this PR: STJ emits `"Pass"`, `"Fail"`, `"Host"`, `"Stun"`, `"TurnUdp"`, `"TurnTcp"`, `"Turns"`.

Any downstream consumer parsing `lshammer scan` stdout (dashboards, alerting, manual scripts) breaks silently — the JSON still parses, the values just no longer match the documented contract.

Fix at the source: turn on the source-gen string-enum converter with a naming policy in `HammerJsonContext`:

```csharp
[JsonSourceGenerationOptions(
    PropertyNamingPolicy = JsonKnownNamingPolicy.CamelCase,
    UseStringEnumConverter = true,
    DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull)]
```

and drop the per-enum `[JsonConverter]` attributes — `UseStringEnumConverter = true` honors the context's naming policy and applies to all enum types in the graph.

Same applies to `ScanTestScenario`.

File: src/FM.LiveSwitch.Hammer/HammerJsonContext.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+using System.Text.Json.Serialization;
+
+namespace FM.LiveSwitch.Hammer
+{
+    [JsonSourceGenerationOptions(
+        PropertyNamingPolicy = JsonKnownNamingPolicy.CamelCase,
+        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull)]
```
Comment (andriy-baranskyy) [inline src/FM.LiveSwitch.Hammer/HammerJsonContext.cs]: Case-sensitivity regression for inbound deserialization.

Newtonsoft.Json was **case-insensitive by default** for property matching. STJ source-gen is **case-sensitive** unless `PropertyNameCaseInsensitive = true` is set on the context. The PR adds `[JsonPropertyName("ipAddresses")]` for one field on `MediaServerInfo`, but every other property on every other DTO (`DeploymentConfig`, `ScanTestMediaServer`, etc.) now matches strictly against `CamelCase`-mapped names.

If any API response field's casing differs from the mapped name (e.g. server emits `publicIpAddress` while `CamelCase` of `PublicIPAddress` is `publicIPAddress`), the field silently deserializes to `null` and you'll see NREs or empty results downstream with no log signal.

Add `PropertyNameCaseInsensitive = true` to the `JsonSourceGenerationOptions` for parity with the old Newtonsoft behavior, or audit every DTO property against the actual wire format and pin with explicit `[JsonPropertyName]`.

File: src/FM.LiveSwitch.Hammer/TrimmerRootDescriptor.xml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+<linker>
+  <assembly fullname="lshammer" preserve="all" />
```
Comment (andriy-baranskyy) [inline src/FM.LiveSwitch.Hammer/TrimmerRootDescriptor.xml]: `<assembly fullname="lshammer" preserve="all" />` preserves the **entire app assembly** — which means trimming is effectively disabled for all of `lshammer`'s own code.

This defeats much of the point of `PublishTrimmed=true` in `publish.yml`. The reason the descriptor exists is that `CommandLineParser` reflects over the verb/option types declared in `lshammer`. But that doesn't require keeping the entire assembly — just the option/verb classes.

Two options:

1. Narrow the descriptor to just the reflection-targeted types:
   ```xml
   <linker>
     <assembly fullname="lshammer">
       <type fullname="FM.LiveSwitch.Hammer.ClusterOptions" preserve="all" />
       <type fullname="FM.LiveSwitch.Hammer.LoadOptions" preserve="all" />
       <type fullname="FM.LiveSwitch.Hammer.ScanOptions" preserve="all" />
     </assembly>
   </linker>
   ```
   The rest of `lshammer` then trims, and the binary shrinks meaningfully.

2. Or accept that `PublishTrimmed` is buying very little for first-party code and document the trade-off in the XML.

With the current descriptor, `PublishTrimmed=true` mostly trims framework assemblies and the small amount of `FM.LiveSwitch` code that isn't reflection-bound — but the savings from trimming the app itself are gone.

**Review by andriy-baranskyy — COMMENTED** [review]
Adversarially-filtered review on `7c88176`. 3 findings survived a 5-agent majority filter (from a 19-candidate toolkit pass).

File: src/FM.LiveSwitch.Hammer/HammerJsonContext.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
+using System.Text.Json.Serialization;
+
+namespace FM.LiveSwitch.Hammer
+{
+    [JsonSourceGenerationOptions(
+        PropertyNamingPolicy = JsonKnownNamingPolicy.CamelCase,
+        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
+        UseStringEnumConverter = true,
+        PropertyNameCaseInsensitive = true)]
```
[the remark this answers is not in the document]
Comment (rustyclarksonfm) [reply src/FM.LiveSwitch.Hammer/HammerJsonContext.cs:9]: Verified against the released binary (current production lshammer, pre-PR). Newtonsoft was already serializing enum values as PascalCase — "state":"Pass", "scenario":"Host" etc. The STJ output matches. No wire format regression. Tested against a live deployment to confirm.

File: src/FM.LiveSwitch.Hammer/MediaServerInfo.cs [hunk 5]
```diff
@@ -18,6 +18,7 @@ class MediaServerInfo
 
         public bool Draining { get; set; }
 
+        [JsonPropertyName("ipAddresses")]
```
Comment (AnkitIsLive) [inline src/FM.LiveSwitch.Hammer/MediaServerInfo.cs:21]: I think property doesn't need's this right since access is case insensitive

Comment (rustyclarksonfm) [reply src/FM.LiveSwitch.Hammer/MediaServerInfo.cs:21]: The attribute ensures serialization matches the previous output — without it, STJ's CamelCase policy produces "iPAddresses" (it only lowercases the first character of a property name, not the full acronym). PropertyNameCaseInsensitive only affects deserialization, not serialization.

**Review by AnkitIsLive — APPROVED** [review]
